### PR TITLE
Adds testing section

### PR DESCRIPTION
**Pull request by villander**
Implements #59 with Unit, Integration and Acceptance tests

### CODE REVIEW

File: markdown/guide/testing.md [hunk 1]
```diff
@@ -1,3 +1,163 @@
 ## Testing Engines
 
-_TODO: PRs welcome! We haven't quite nailed down the testing story for Engines and especially Lazy Engines, so if you have ideas reach out!_
+> in-repo-engine & in-repo-addon doesn't have support to separate tests. If you want to create a separation of concerns between the app and your engine, to extract this into a normal addon.
```
Comment (jenweber) [inline markdown/guide/testing.md]:
```suggestion
> The in-repo-engine & in-repo-addon don't have support for separate tests. If you want to create a separation of concerns between the app and your engine, extract the functionality into a normal addon.
```

File: markdown/guide/testing.md [hunk 2]
```diff
@@ -1,3 +1,163 @@
...
+
+* for test an `in-repo-engine` we must setup the tests from the host app
```
Comment (jenweber) [inline markdown/guide/testing.md]:
```suggestion
* To test an `in-repo-engine`, we must set up the tests from the host app
```

File: markdown/guide/testing.md [hunk 3]
```diff
@@ -1,3 +1,163 @@
...
+
+* for test an `engine as addon` we must setup the tests from the dummy app
```
Comment (jenweber) [inline markdown/guide/testing.md]:
```suggestion
* To test an `engine as an addon`, we must set up the tests from the dummy app
```

File: markdown/guide/testing.md [hunk 4]
```diff
@@ -1,3 +1,163 @@
...
+
+Ember Engines comes with a **set of test helpers that can be used in Unit/Integration tests**:
+
+`engineResolverFor` gets the resolver class used by an Engine and creates an instance to be used with test modules
```
Comment (jenweber) [inline markdown/guide/testing.md]:
```suggestion
`engineResolverFor` gets the resolver class used by an Engine and creates an instance to be used with test modules:
```

File: markdown/guide/testing.md [hunk 5]
```diff
@@ -1,3 +1,163 @@
...
+// tests/<integration or unit>/…
+import engineResolverFor from 'ember-engines/test-support/engine-resolver-for';
+```
+
+### Unit/Integration Testing
+
+
+To test candidates for unit/integration (e.g. components, services and controllers) declared inside an engine, you need to set a custom resolver with the engine's prefix using `engineResolverFor` helper.
+
+Assuming that we have an `in-repo-engine or engine as addon` called `admin-app` and it has a component `hello-name` and service `some-thing`. The following would be the setup to test such component from the `<app-name>` (host app or dummy app):
```
Comment (jenweber) [inline markdown/guide/testing.md]:
```suggestion
What does it look like to test a component from a host app or dummy app? Let's go over some examples in the next section. In the following tests, `admin-app` is an `in-repo-engine` or `engine as addon`, `hello-name` is a component, and `some-thing` is a service.
```

File: markdown/guide/testing.md [hunk 6]
```diff
@@ -1,3 +1,163 @@
...
+
+#### Unit Testing Basics
+
+Supposing that we have in the engine a service that has a `computedFoo` computed property based on a `foo` property.
```
Comment (jenweber) [inline markdown/guide/testing.md]:
```suggestion
In the first scenario, suppose that the engine has a service with a `computedFoo` computed property, based on a `foo` property:
```

File: markdown/guide/testing.md [hunk 7]
```diff
@@ -1,3 +1,163 @@
...
+export default Service.extend({
+  foo: 'bar',
+
+  computedFoo: computed('foo', function() {
+    return `computed ${this.foo}`;
+  })
+});
+```
+
+The unit test will be like this:
```
Comment (jenweber) [inline markdown/guide/testing.md:43]:
```suggestion
Here is how we could unit test the service:
```

File: markdown/guide/testing.md [hunk 8]
```diff
@@ -1,3 +1,163 @@
...
+    someThing.set('foo', 'baz');
+
+    assert.equal(someThing.get('computedFoo'), 'computed baz');
+  });
+});
+```
+
+#### Testing Components
+
+Supposing that we have in the engine this component
```
Comment (jenweber) [inline markdown/guide/testing.md]:
```suggestion
Next, suppose that our engine has a component:
```

File: markdown/guide/testing.md [hunk 9]
```diff
@@ -1,3 +1,163 @@
...
+
+```hbs
+{{!--  admin-app/addon/components/hello-name.hbs --}}
+Hello, {{name}}!
+```
+
+The integration test will be like this:
```
Comment (jenweber) [inline markdown/guide/testing.md]:
```suggestion
Here is what the integration test for that component might look:
```

File: markdown/guide/testing.md [hunk 10]
```diff
@@ -1,3 +1,163 @@
...
+  test('it renders', async function(assert) {
+
+    await render(hbs`{{#hello-name name="Tom"}}{{/hello-name}}`);
+
+    assert.equal(this.element.textContent.trim(), 'Hello, Tom!');
+  });
+});
+```
+
+**Note: you could create a helper and then use it like `Resolver from ../helpers/admin-app/resolver`**
```
Comment (jenweber) [inline markdown/guide/testing.md]: I'm not sure what you mean by this note. Can you add some more detail here?

File: markdown/guide/testing.md [hunk 11]
```diff
@@ -1,3 +1,163 @@
...
+
+
+#### Acceptance Testing
+
+Supposing that we are mouting `admin-app` on host-app router
```
Comment (jenweber) [inline markdown/guide/testing.md]:
```suggestion
Suppose that we are mouting `admin-app` on host-app router:
```

File: markdown/guide/testing.md [hunk 12]
```diff
@@ -1,3 +1,163 @@
...
+});
+
+Router.map(function() {
+  this.mount('admin');
+});
+
+export default Router;
+```
+
+The acceptance test will be like this:
```
Comment (jenweber) [inline markdown/guide/testing.md]: Here is an acceptance test for routing:

**Review by jenweber — CHANGES_REQUESTED** [review]

**Comment by jenweber** [comment]
Thanks for writing this! Just some wording changes suggested.

File: markdown/guide/testing.md (hunk 10 above)
Comment (villander) [reply markdown/guide/testing.md]: fixed

File: markdown/guide/testing.md (hunk 12 above)
Comment (villander) [reply markdown/guide/testing.md]: fixed!

File: markdown/guide/testing.md [hunk 13]
```diff
@@ -1,3 +1,160 @@
...
+  test('the user can visit /admin page', async function(assert) {
+    await visit('/');
+    await click('.admin-menu-item');
+
+    assert.equal(currentURL(), '/admin');
+  });
+});
+```
+
+### Testing for Lazy Engines
```
Comment (villander) [inline markdown/guide/testing.md]: @rwjblue

**Comment by villander** [comment]
@dgeb can we merge this? this will be a fullpower for users of ember engines.

File: markdown/guide/testing.md [hunk 14]
```diff
@@ -1,3 +1,160 @@
 ## Testing Engines
 
-_TODO: PRs welcome! We haven't quite nailed down the testing story for Engines and especially Lazy Engines, so if you have ideas reach out!_
+> The in-repo-engine & in-repo-addon don't have support for separate tests. If you want to create a separation of concerns between the app and your engine, extract the functionality into a normal addon.
```
Comment (dgeb) [inline markdown/guide/testing.md]:
```suggestion
> The in-repo-engine & in-repo-addon don't have support for separate tests. If you want to maintain a separation of concerns between your app and engine, convert the engine into a separate addon.
```

File: markdown/guide/testing.md [hunk 15]
```diff
@@ -1,3 +1,160 @@
...
+### Unit/Integration Testing
+
+
+To test candidates for unit/integration (e.g. components, services and controllers) declared inside an engine, you need to set a custom resolver with the engine's prefix using `engineResolverFor` helper.
+
+What does it look like to test a component from a host app or dummy app? Let's go over some examples in the next section. In the following tests, `admin-app` is an `in-repo-engine` or `engine as addon`, `hello-name` is a component, and `some-thing` is a service.
+
+#### Unit Testing Basics
+
+Supposing that we have in the engine a service that has a `computedFoo` computed property based on a `foo` property.
```
Comment (dgeb) [inline markdown/guide/testing.md]:
```suggestion
Suppose that we have in the engine a service that has a `computedFoo` computed property based on a `foo` property.
```

File: markdown/guide/testing.md [hunk 16]
```diff
@@ -1,3 +1,160 @@
...
+// tests/<integration or unit>/…
+import engineResolverFor from 'ember-engines/test-support/engine-resolver-for';
+```
+
+### Unit/Integration Testing
+
+
+To test candidates for unit/integration (e.g. components, services and controllers) declared inside an engine, you need to set a custom resolver with the engine's prefix using `engineResolverFor` helper.
+
+What does it look like to test a component from a host app or dummy app? Let's go over some examples in the next section. In the following tests, `admin-app` is an `in-repo-engine` or `engine as addon`, `hello-name` is a component, and `some-thing` is a service.
```
Comment (dgeb) [inline markdown/guide/testing.md]: I think this is very confusing to call our engine `admin-app`, which makes it sound like an "app". How about `admin-console` or `admin-section` or even `admin-engine`?

Comment (dgeb) [reply markdown/guide/testing.md]: Also, let's just use a single term `in-repo-engine` - saying `or 'engine as addon'` is super confusing.

File: markdown/guide/testing.md [hunk 17]
```diff
@@ -1,3 +1,160 @@
...
+#### Testing Components
+
+Next, suppose that our engine has a component:
+
+```hbs
+{{!--  admin-app/addon/components/hello-name.hbs --}}
+Hello, {{name}}!
+```
+
+Here is what the integration test for that component might look:
```
Comment (dgeb) [inline markdown/guide/testing.md]:
```suggestion
Here's an example integration test for that component:
```

File: markdown/guide/testing.md [hunk 18]
```diff
@@ -1,3 +1,160 @@
...
+});
+
+Router.map(function() {
+  this.mount('admin');
+});
+
+export default Router;
+```
+
+Here is an acceptance test for routing::
```
Comment (dgeb) [inline markdown/guide/testing.md]:
```suggestion
Here is an acceptance test for routing:
```

File: markdown/guide/testing.md [hunk 19]
```diff
@@ -1,3 +1,160 @@
...
+    await click('.admin-menu-item');
+
+    assert.equal(currentURL(), '/admin');
+  });
+});
+```
+
+### Testing for Lazy Engines
+
+If you have a lazy engine, you'll need to edit your `tests/test-helper.js` like this:
```
Comment (dgeb) [inline markdown/guide/testing.md]:
```suggestion
If you have a lazy engine, you'll need to ensure that your `tests/test-helper.js` is configured to preload your engine's assets:
```

File: markdown/guide/testing.md [hunk 20]
```diff
@@ -1,3 +1,160 @@
...
+import { start } from 'ember-qunit';
+import preloadAssets from 'ember-asset-loader/test-support/preload-assets';
+import manifest from '<app-name>/config/asset-manifest';
+
+setApplication(Application.create(config.APP));
+
+preloadAssets(manifest).then(start); // This ensures all engine resources are loaded before the tests
+```
+
+This should be enough to make unit, integration & acceptance tests work.
```
Comment (dgeb) [inline markdown/guide/testing.md]:
```suggestion
This should be sufficient to make unit, integration, and acceptance tests work.
```

**Review by dgeb — CHANGES_REQUESTED** [review]

File: markdown/guide/testing.md (hunk 16 above)
Comment (villander) [reply markdown/guide/testing.md]: @dgeb how can we referrer here? `in-repo-engine` and `addon-engine`?

Comment (dgeb) [reply markdown/guide/testing.md]: > `in-repo-engine` and `addon-engine`

Sure that's fine to distinguish between the two flavors of engines

My objection is that this phrase doesn't make sense:

> `admin-app` is an `in-repo-engine` or `engine as addon`

It has to be one or the other.

Comment (villander) [reply markdown/guide/testing.md]: fixed!

File: markdown/guide/testing.md [hunk 21]
```diff
@@ -1,3 +1,160 @@
 ## Testing Engines
 
-_TODO: PRs welcome! We haven't quite nailed down the testing story for Engines and especially Lazy Engines, so if you have ideas reach out!_
+> The in-repo-engine & in-repo-addon don't have support for separate tests. If you want to maintain a separation of concerns between your app and engine, convert the engine into a separate addon.
+
+* To test an `in-repo-engine`, we must set up the tests from the host app
```
Comment (dgeb) [inline markdown/guide/testing.md]:
```suggestion
* To test an in-repo engine, co-locate the engine's tests with the host app's tests, just like any other in-repo addon.
```

File: markdown/guide/testing.md [hunk 22]
```diff
@@ -1,3 +1,160 @@
...
+
+* To test an `engine as an addon`, we must set up the tests from the dummy app
```
Comment (dgeb) [inline markdown/guide/testing.md]:
```suggestion
* To test a standalone engine, use the dummy app, just like any other standalone addon project.
```

File: markdown/guide/testing.md [hunk 23]
```diff
@@ -1,3 +1,160 @@
 ## Testing Engines
 
-_TODO: PRs welcome! We haven't quite nailed down the testing story for Engines and especially Lazy Engines, so if you have ideas reach out!_
+> The in-repo-engine & in-repo-addon don't have support for separate tests. If you want to maintain a separation of concerns between your app and engine, convert the engine into a separate addon.
```
Comment (dgeb) [inline markdown/guide/testing.md]:
```suggestion
Just like other addons, engines must be tested differently depending upon whether they are "in-repo" or "standalone":
```

**Review by dgeb — CHANGES_REQUESTED** [review]
I've left some suggestions that attempt to clarify the difference between project types.

**Comment by villander** [comment]
thanks for suggestions! Applied

**Review by dgeb — APPROVED** [review]

**Comment by dgeb** [comment]
Thank _you_ @villander! This section should be very helpful.

**Pull request merged**